### ISSUE TITLE
Incorrect midpoint if interval contains 0

### ISSUE THREAD

**Issue by freemin7** [JuliaMath/Roots.jl#271]
`__middle(Float16, UInt16, Float16(-1.5), Float16(1.4) )` results in `Float16(-1.45)`. However counting reveals that this midpoint is far of from the actual midpoint.

```
a, b = Float16(-1.5),Float16(-1.5)
while a < 1.4
       a = nextfloat(nextfloat(a))
       b = nextfloat(b)
end
```
After the loop `a` is `Float16(1.4)` while `b` is `Float16(-3.04e-6)`.

**Comment by jverzani** [comment]
The expectation for `__middle` is to have both `x` and `y` have the same sign (aside from -0.0 and 0.0 differences) so the 1.45 makes sense, as it is the middle of 1.4 and 1.5. Where the algorithm is inefficient (but does strip of a bit, so it does work) is near `0`:

```
julia> Roots.__middle(Float16(0.0), Float16(1.0))
Float16(0.00586)

julia> Roots.__middle(Float32(0.0), Float32(1.0))
8.131516f-20

julia> Roots.__middle(Float64(0.0), Float64(1.0))
1.118751109680031e-154
```

This bit of code could use some cleaning up though!  Your issues showed me update state was calling _middle and not __middle as it should!

**Comment by jverzani** [comment]
Closed by #272. Reopen if this isn't sufficient.
